pkg/util/git.go

```diff
@@ -0,0 +1,276 @@
+
+package util
+
+import (
+	"fmt"
+	gitpkg "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	githttp "github.com/go-git/go-git/v5/plumbing/transport/http"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	GitHubHost    string = "github.com"
+	RawGitHubHost string = "raw.githubusercontent.com"
+	GitLabHost    string = "gitlab.com"
+	BitbucketHost string = "bitbucket.org"
+
+	GitHubToken    string = "GITHUB_TOKEN"
+	GitLabToken    string = "GITLAB_TOKEN"
+	BitbucketToken string = "BITBUCKET_TOKEN"
+)
+
+type GitUrl struct {
+	Protocol string
+	Host     string
+	Owner    string
+	Repo     string
+	Branch   string
+	Path     string
+	token    string
+	IsFile   bool
+}
+
+// ParseGitUrl extracts information from a GitHub, GitLab, or Bitbucket url
+// A client is used to check whether the url is private or public, and sets
+// the providers personal access token from the environment variable
+func ParseGitUrl(fullUrl string) (GitUrl, error) {
+	var c = http.Client{
+		Timeout: HTTPRequestResponseTimeout,
+	}
+	return parseGitUrlWithClient(fullUrl, c)
+}
+
+func parseGitUrlWithClient(fullUrl string, c http.Client) (GitUrl, error) {
+	var g GitUrl
+
+	err := ValidateURL(fullUrl)
+	if err != nil {
+		return g, err
+	}
+
+	parsedUrl, err := url.Parse(fullUrl)
+	if err != nil {
+		return g, err
+	}
+
+	if len(parsedUrl.Path) == 0 {
+		return g, fmt.Errorf("url path should not be empty")
+	}
+
+	if parsedUrl.Host == RawGitHubHost || parsedUrl.Host == GitHubHost {
+		g, err = parseGitHubUrl(g, parsedUrl, c)
+	} else if parsedUrl.Host == GitLabHost {
+		g, err = parseGitLabUrl(g, parsedUrl, c)
+	} else if parsedUrl.Host == BitbucketHost {
+		g, err = parseBitbucketUrl(g, parsedUrl, c)
+	} else {
+		err = fmt.Errorf("url host should be a valid GitHub, GitLab, or Bitbucket host; received: %s", parsedUrl.Host)
+	}
+
+	return g, err
+}
+
+func parseGitHubUrl(g GitUrl, url *url.URL, c http.Client) (GitUrl, error) {
+	var splitUrl []string
+	var err error
+
+	g.Protocol = url.Scheme
+	g.Host = url.Host
+
+	if g.Host == RawGitHubHost {
+		g.IsFile = true
+		// raw GitHub urls don't contain "blob" or "tree"
+		splitUrl = strings.SplitN(url.Path[1:], "/", 4)
+		if len(splitUrl) == 4 {
+			g.Owner = splitUrl[0]
+			g.Repo = splitUrl[1]
+			g.Branch = splitUrl[2]
+			g.Path = splitUrl[3]
+		} else {
+			err = fmt.Errorf("raw url path should contain <owner>/<repo>/<branch>/<path/to/file>, received: %s", url.Path[1:])
+		}
+	}
+
+	if g.Host == GitHubHost {
+		splitUrl = strings.SplitN(url.Path[1:], "/", 5)
+		if len(splitUrl) < 2 {
+			err = fmt.Errorf("url path should contain <user>/<repo>, received: %s", url.Path[1:])
+		} else {
+			g.Owner = splitUrl[0]
+			g.Repo = splitUrl[1]
+
+			if len(splitUrl) == 5 {
+				switch splitUrl[2] {
+				case "tree":
+					g.IsFile = false
+				case "blob":
+					g.IsFile = true
+				}
+				g.Branch = splitUrl[3]
+				g.Path = splitUrl[4]
+			}
+		}
+	}
+
+	if !isGitUrlPublic(g, c) {
```

Comment:
We shouldn't rely on a false value to assume it's a private repo.  False can mean the URL is unreachable/non-existent.  We also need to consider error scenarios where users may be providing a private URL but have given us a bad token or did not set the env var.   

- If the token does not exist, we should issue an error asking users to either check their URL or set the token if private repo was intended.  

- If the token exists, we should check if it's reachable.  If not, issue an error and stop the processing

Reply:
Not sure if parsing the url should make http calls, so I moved the token check to the cloning function